### PR TITLE
chore: Webhook event data improvements

### LINKED ISSUE #4180: Inspect the conversation-update event, include changed parameter in the payload
The current webhook event does not contain the information about which attribute has changed. It would be helpful to send the present and previous values of the changed params in the payload. This would help to build custom workflow on top of param changes.

### PR DESCRIPTION
**Pull request by tejaswinichile**
# Pull Request Template

## Description

Added changed attributes in the payload for webhook.

Fixes #4180 

## Type of change

Please delete options that are not relevant.

- [ ] New feature (non-breaking change which adds functionality)

## How Has This Been Tested?

Tested on local

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [ ] I have commented on my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

```
{
   "additional_attributes": {
      "browser": {
         "device_name": "Unknown",
         "browser_name": "Chrome",
         "platform_name": "macOS",
         "browser_version": "97.0.4692.71",
         "platform_version": "10.14.6"
      },
      "referer": "http://localhost:3000/widget_tests",
      "initiated_at": {
         "timestamp": "Sat Jan 22 2022 13:27:05 GMT+0530 (India Standard Time)"
      }
   },
   "can_reply": true,
   "channel": "Channel::WebWidget",
   "contact_inbox": {
      "ContactInbox": "0x00007fd280ad2f50",
      "id": 2,
      "contact_id": 1,
      "inbox_id": 1,
      "source_id": "af156dd9-74cb-4131-9a3e-9ea800bcae05",
      "created_at": "Mon, 16 Aug 2021 13:40:12.462364000 UTC +00:00",
      "updated_at": "Mon, 31 Jan 2022 14:08:51.080299000 UTC +00:00",
      "hmac_verified": true,
      "pubsub_token": "SpgznQD8YCv9mjLJCocNfjR3"
   },
   "id": 62,
   "inbox_id": 1,
   "messages": [
      {
         "id": 1230,
         "content": "help please\n\n111",
         "account_id": 1,
         "inbox_id": 1,
         "conversation_id": [
            62,
            62
         ],
         "message_type": [
            "incoming",
            0
         ],
         "created_at": [
            "Wed, 23 Mar 2022 09:17:44.550419000 UTC +00:00",
            1648027064
         ],
         "updated_at": "Wed, 23 Mar 2022 09:17:45.848698000 UTC +00:00",
         "private": false,
         "status": "sent",
         "source_id": null,
         "content_type": "text",
         "content_attributes": {},
         "sender_type": "Contact",
         "sender_id": 1,
         "external_source_ids": {
            "slack": "cw-origin-1648027104.714449"
         },
         "conversation": {
            "assignee_id": 1
         },
         "sender": {
            "additional_attributes": {
               "description": "",
               "company_name": "",
               "social_profiles": {
                  "github": "",
                  "twitter": "",
                  "facebook": "",
                  "linkedin": ""
               }
            },
            "custom_attributes": {
               "priority_number": "5",
               "is_priority_customer": true
            },
            "email": "jane@example.com",
            "id": 1,
            "identifier": null,
            "name": "jane",
            "phone_number": "+2320000",
            "pubsub_token": "yMhuucfqujTtNCjLbwdu8Tid",
            "thumbnail": "https://www.gravatar.com/avatar/9e26471d35a78862c17e467d87cddedf?d=404",
            "type": "contact"
         }
      }
   ],
   "meta": {
      "sender": {
         "additional_attributes": {
            "description": "",
            "company_name": "",
            "social_profiles": {
               "github": "",
               "twitter": "",
               "facebook": "",
               "linkedin": ""
            }
         },
         "custom_attributes": {
            "priority_number": "5",
            "is_priority_customer": true
         },
         "email": "jane@example.com",
         "id": 1,
         "identifier": null,
         "name": "jane",
         "phone_number": "+2320000",
         "pubsub_token": "yMhuucfqujTtNCjLbwdu8Tid",
         "thumbnail": "https://www.gravatar.com/avatar/9e26471d35a78862c17e467d87cddedf?d=404",
         "type": "contact"
      },
      "assignee": {
         "id": 1,
         "name": "John",
         "available_name": "John",
         "avatar_url": "http://0.0.0.0:3000/rails/active_storage/representations/redirect/eyJfcmFpbHMiOnsibWVzc2FnZSI6IkJBaHBBaHdCIiwiZXhwIjpudWxsLCJwdXIiOiJibG9iX2lkIn19--2ace74960566b4cf4316f29bdee82485adfa1081/eyJfcmFpbHMiOnsibWVzc2FnZSI6IkJBaDdCem9MWm05eWJXRjBTU0lJYW5CbkJqb0dSVlE2QzNKbGMybDZaVWtpRERJMU1IZ3lOVEFHT3daVSIsImV4cCI6bnVsbCwicHVyIjoidmFyaWF0aW9uIn19--76aa1ec8cca51dc93908668e46f2f6729fbbd445/1610894557728.jpg",
         "type": "user",
         "availability_status": null,
         "thumbnail": "http://0.0.0.0:3000/rails/active_storage/representations/redirect/eyJfcmFpbHMiOnsibWVzc2FnZSI6IkJBaHBBaHdCIiwiZXhwIjpudWxsLCJwdXIiOiJibG9iX2lkIn19--2ace74960566b4cf4316f29bdee82485adfa1081/eyJfcmFpbHMiOnsibWVzc2FnZSI6IkJBaDdCem9MWm05eWJXRjBTU0lJYW5CbkJqb0dSVlE2QzNKbGMybDZaVWtpRERJMU1IZ3lOVEFHT3daVSIsImV4cCI6bnVsbCwicHVyIjoidmFyaWF0aW9uIn19--76aa1ec8cca51dc93908668e46f2f6729fbbd445/1610894557728.jpg"
      },
      "hmac_verified": true
   },
   "status": "open",
   "custom_attributes": {},
   "snoozed_until": null,
   "unread_count": 0,
   "agent_last_seen_at": 1648472816,
   "contact_last_seen_at": 1648026830,
   "timestamp": 1648472186,
   "event": "conversation_updated",
   "changed_attributes": [
      {
         "status": [
            "open",
            "resolved"
         ],
         "updated_at": [
            "Mon, 28 Mar 2022 12:56:26.350033000 UTC +00:00, Mon, 28 Mar 2022 12:56:49.267135000 UTC +00:00"
         ]
      }
   ]
}
```

### CODE REVIEW

**Comment by tejaswinichile** [comment]
```
 "changed_attributes": [
      {
         "status": {
            previous_value: "open",
            current_value: "resolved"
         }
      }
   ]

File: app/listeners/base_listener.rb [hunk 1]
```diff
@@ -22,4 +22,10 @@ def extract_contact_and_account(event)
     contact = event.data[:contact]
     [contact, contact.account]
   end
+
+  def extract_changed_attributes(event)
+    changed_attributes = event.data[:changed_attributes]
+    changed_attributes.map { |k, v| { k => { previous_value: v[0], current_value: v[1] } } }
```
Comment (tejaswinichile) [inline app/listeners/base_listener.rb:28]: Made this change to handle https://github.com/chatwoot/chatwoot/pull/4317#issuecomment-1081463383

**Review by sojan-official — COMMENTED** [review]
Can we also have a spec for webhook listener to ensure we don't end up breaking the webhook format ?

**Comment by tejaswinichile** [comment]
@sojan-official Oh sure, I just fixed previous specs, let me add new specs for the same.
As per the old case, we would have empty changed_attributes if changes were not available.

**Comment by tejaswinichile** [comment]
@sojan-official I have added the specs for the webhook listener.

**Review by sojan-official — APPROVED** [review]

**Pull request merged**